pkg/backoff/backoff.go

```diff
@@ -0,0 +1,115 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package backoff
+
+import (
+	"math"
+	"math/rand"
+	"time"
+
+	"github.com/pingcap/dm/pkg/terror"
+)
+
+// Backoff is an exponential counter, it starts from `Min` duration, and after
+// every call to `Duration` method the duration will be multiplied by `Factor`,
+// but it never exceeds `Max`. Backoff is not thread-safe
+type Backoff struct {
+	// cwnd is the congestion window
+	cwnd int
+
+	// Factor is the multiplying factor for each increment step
+	Factor float64
+
+	// Jitter eases contention by randomizing backoff steps
+	Jitter bool
+
+	// Min and Max are the minimum and maximum values of the counter
+	Min, Max time.Duration
+}
+
+// NewBackoff creates a new backoff instance
+func NewBackoff(factor float64, jitter bool, min, max time.Duration) (*Backoff, error) {
+	if factor <= 0 {
+		return nil, terror.ErrBackoffArgsNotValid.Generate("factor", factor)
+	}
+	if min < 0 {
+		return nil, terror.ErrBackoffArgsNotValid.Generate("min", min)
+	}
+	if max < 0 || max < min {
+		return nil, terror.ErrBackoffArgsNotValid.Generate("max", max)
+	}
+	return &Backoff{
+		Factor: factor,
+		Jitter: jitter,
+		Min:    min,
+		Max:    max,
+	}, nil
+}
+
+// Duration returns the duration for the current cwnd and increases the cwnd counter
+func (b *Backoff) Duration() time.Duration {
```

Comment:
this function only used in test?  if so, can we remove this function, or change it to private

Reply:
it is common usage